File: src/model/KNN.py
import csv
import Model
from Model import Model
from statistics import mode

class KNearestNeighbor(Model):
    def score(self, params = None):
        """
        Scores performance based on KNN results to the test dataset

        returns: a score element of accuracy 
        """
        tp = 0
        k = 3
        for i in range(len(self.data_test_x)):
            result = self.KNN(k, self.data_test_x[i])
            if(result == self.data_test_y[i][0]):
                tp += 1
    
        return (tp/len(self.data_test_x))  

    def distance(self, query, row):
        """
        Calculates the distance (difference) between the query (a row from the test dataset) and the row from train dataset

        returns: the distance in a form of integer
        """
        dist = 0
        for i in range(len(query)):
            if query[i] != row[i]:
                dist += 1
        return dist

    def KNN(self, k, query):
        """
        Calculates the classification of a single query (row from test dataset) using the KNN algorithm

        returns: the result classification
        """
        dist_array = [[0, i] for i in range(len(self.data_train_x))]
        for i in range(len(self.data_train_x)):
            dist_array[i][0] = self.distance(query, self.data_train_x[i])
        
        sorted_dist = sorted(dist_array, key=lambda x: x[0])
        sorted_dist = sorted_dist[:k]
        result = [self.data_train_y[dist[1]][0] for dist in sorted_dist]
        return mode(result)


if __name__ == "__main__":
    with open('../data/dummy.csv', newline='') as csvfile:
        reader = csv.reader(csvfile)
        data_csv = [row for row in reader]
        data_csv.pop(0)

    kNN = KNearestNeighbor(data_csv[:7], data_csv[7:], len(data_csv[0]) - 1)
    print(kNN.score())

File: src/model/KNN.go
package model

import (
	"fmt"
	"sort"
	"sync"
)

type KNearestNeighbor struct {
	dataTrainX [][]string
	dataTrainY [][]string
	dataTestX  [][]string
	dataTestY  [][]string
}

func (knn *KNearestNeighbor) distance(query []string, row []string) int {
	// var wg sync.WaitGroup
	// var mu sync.Mutex

	dist := 0
	// for i := range query {
	// 	wg.Add(1)
	// 	go func(i int) {
	// 		defer wg.Done()
	// 		if query[i] != row[i] {
	// 			mu.Lock()
	// 			dist++
	// 			mu.Unlock()
	// 		}
	// 	}(i)
	// }

	// wg.Wait()

	for i := range query {
		if query[i] != row[i] {
			dist++
		}
	}

	return dist
}


func (knn *KNearestNeighbor) predict(k int, query []string) string {
	distArray := make([][2]int, len(knn.dataTrainX))

	for i := range knn.dataTrainX {
		distArray[i] = [2]int{knn.distance(query, knn.dataTrainX[i]), i}
	}

	sort.Slice(distArray, func(i, j int) bool {
		if distArray[i][0] == distArray[j][0] {
			return distArray[i][1] < distArray[j][1]
		}
		return distArray[i][0] < distArray[j][0]
	})

	// fmt.Println(distArray)
	distArray = distArray[:k]

	result := make(map[string]int)
	for _, pair := range distArray {
		label := knn.dataTrainY[pair[1]][0]
		result[label]++
	}

	maxCount := 0
	mode := ""
	for label, count := range result {
		if count > maxCount {
			maxCount = count
			mode = label
		}
	}

	// fmt.Println(result)
	return mode
}

func (knn *KNearestNeighbor) score(k int) float64 {
	var wg sync.WaitGroup
	var mu sync.Mutex

	tp := 0

	for i, query := range knn.dataTestX {
		wg.Add(1)
		go func(i int, query []string) {
			defer wg.Done()

			prediction := knn.predict(k, query)
			if prediction == knn.dataTestY[i][0] {
				mu.Lock() // Lock tp variable
				tp++
				mu.Unlock() // Unlock after modifying tp
			}
			// fmt.Println(i)
			// fmt.Printf("Current Accuracy: %.2f\n", float64(tp)/float64(i+1))
		}(i, query)
	}
	wg.Wait()

	return float64(tp) / float64(len(knn.dataTestX))
}

func MainKNN(data [][]string) {

	var features [][]string
	var labels [][]string
	for _, row := range data {
		features = append(features, row[:len(row)-1])
		labels = append(labels, []string{row[len(row)-1]})
	}

	// trainX := features
	// testX := features
	// trainY := labels
	// testY := labels
	trainX, testX := splitDataset(features, 0.8)
	trainY, testY := splitDataset(labels, 0.8)

	kNN := KNearestNeighbor{
		dataTrainX: trainX,
		dataTrainY: trainY,
		dataTestX:  testX,
		dataTestY:  testY,
	}
	score := kNN.score(3)
	fmt.Printf("Accuracy: %.2f\n", score)
}
